### PR TITLE
Project controller

### PR DESCRIPTION
**Pull request by End-rey**
I add get request for all projects and for 1 project by id, and post request for project

### CODE REVIEW

File: src/main/java/org/spburegistry/backend/ExceptionHandler/GlobalExceptionHandler.java [hunk 1]
```diff
@@ -5,15 +5,15 @@
...
 
 @ControllerAdvice
 @Slf4j
 public class GlobalExceptionHandler {
     @ExceptionHandler
-    public ResponseEntity<AppError> catchNoSuchUserException(NoSuchUserException e) {
+    public ResponseEntity<AppError> catchNoSuchUserException(NoSuchEntityException e) {
         log.error(e.getMessage(), e);
         return new ResponseEntity<>(new AppError(HttpStatus.NOT_FOUND.value(), e.getMessage()), HttpStatus.NOT_FOUND);
     }
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/ExceptionHandler/GlobalExceptionHandler.java:19]: I think it is needed to add abstract exception handler to handle any unexpected errors

File: src/main/java/org/spburegistry/backend/controller/UserController.java [hunk 2]
```diff
@@ -13,12 +15,12 @@ public class UserController {
...
 
     @GetMapping("/users/user")
-    public User getUser(@RequestParam() Long id) {
+    public UserTO getUser(@RequestParam() Long id) {
         return userService.findById(id);
     }
 
     @GetMapping("/users")
-    public Iterable<User> getAllUsers() {
+    public Set<UserTO> getAllUsers() {
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/controller/UserController.java:23]: Why did you change Iterable to Set?

File: src/main/java/org/spburegistry/backend/service/UserService.java [hunk 3]
```diff
@@ -1,9 +1,10 @@
...
-import org.spburegistry.backend.entity.User;
+import java.util.Set;
+
+import org.spburegistry.backend.dto.UserTO;
 
 public interface UserService {
-    public Iterable<User> findAll();
-    public User findById(long id);
-    public User findByName(String name);
+    public Set<UserTO> findAll();
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/service/UserService.java:8]: I think it is better to leave Iterable for interface declaration because it it more flexible than Set

File: src/main/java/org/spburegistry/backend/dto/ProjectTO.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    private String requirements;
+
+    private ClientTO client;
+
+    private Set<StudentTO> students;
+
+    private Set<ClinicTO> clinics;
+
+    private Set<String> tags;
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/dto/ProjectTO.java:32]: I think it is better to add TagTO and find tags by id

File: src/main/java/org/spburegistry/backend/entity/Clinic.java [hunk 5]
```diff
@@ -17,12 +18,19 @@ public class Clinic extends BaseEntity {
...
 
+    @EqualsAndHashCode.Exclude
+    @Builder.Default
     @ManyToMany(mappedBy = "clinics")
-    private Set<Project> projects;
+    private Set<Project> projects = new HashSet<>();
+
+    public void addProject(Project project) {
+        projects.add(project);
+    }
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/entity/Clinic.java:33]: We don't need this method. Binding will be done by adding clinics to project.

File: src/main/java/org/spburegistry/backend/entity/Project.java [hunk 6]
```diff
@@ -6,87 +6,98 @@
...
+                name = "project_student", 
+                joinColumns = @JoinColumn(name = "project_id"), 
+                inverseJoinColumns = @JoinColumn(name = "student_id")
+        )
+        private Set<Student> students = new HashSet<>();
+
+        public void addStudent(Student student) {
+                students.add(student);
+                student.addProject(this);
+        }
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/entity/Project.java:68]: Unused method.

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        Set<CommitTO> commits = project.getCommits().stream()
+                .map(commit -> commitToTO(commit))
+                .collect(Collectors.toSet());
+
+        Set<String> links = project.getLinks().stream()
+                .map(link -> link.getLink())
+                .collect(Collectors.toSet());
+
+        Set<String> requirementsForPerformers = project.getRequirementsForPerformers().stream()
+                .map(re -> re.getRequirement())
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:100]:
```suggestion
                .map(Requirement::getRequirement)
```

**Review by SerjVankovich — CHANGES_REQUESTED** [review]

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+import org.spburegistry.backend.dto.StudentTO;
+import org.spburegistry.backend.dto.UserTO;
+import org.spburegistry.backend.entity.Client;
+import org.spburegistry.backend.entity.Clinic;
+import org.spburegistry.backend.entity.Commit;
+import org.spburegistry.backend.entity.Project;
+import org.spburegistry.backend.entity.Student;
+import org.spburegistry.backend.entity.User;
+
+public class ConvertToTO {
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:21]: Please create private constructor for this class

File: src/main/java/org/spburegistry/backend/ExceptionHandler/GlobalExceptionHandler.java [hunk 9]
```diff
@@ -5,15 +5,15 @@
...
+import org.spburegistry.backend.ExceptionHandler.exception.NoSuchEntityException;
 
 import lombok.extern.slf4j.Slf4j;
 
 @ControllerAdvice
 @Slf4j
 public class GlobalExceptionHandler {
     @ExceptionHandler
-    public ResponseEntity<AppError> catchNoSuchUserException(NoSuchUserException e) {
+    public ResponseEntity<AppError> catchNoSuchUserException(NoSuchEntityException e) {
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/ExceptionHandler/GlobalExceptionHandler.java:16]: Done

File: src/main/java/org/spburegistry/backend/ExceptionHandler/GlobalExceptionHandler.java (hunk 1 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/ExceptionHandler/GlobalExceptionHandler.java:19]: Done

File: src/main/java/org/spburegistry/backend/service/implementation/UserServiceImpl.java [hunk 10]
```diff
@@ -18,19 +23,17 @@ public UserServiceImpl(UserRepo userRepo) {
...
+            () -> new NoSuchEntityException("User with id " + id + " not found"));
+        return ConvertToTO.userToTO(user);
     }
 
     @Override
-    public Iterable<User> findAll() {
-        return userRepo.findAll();
+    public Set<UserTO> findAll() {
+        return userRepo.findAll().stream()
+            .map(user -> ConvertToTO.userToTO(user))
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/implementation/UserServiceImpl.java:35]: Done

File: src/main/java/org/spburegistry/backend/service/implementation/UserServiceImpl.java [hunk 11]
```diff
@@ -18,19 +23,17 @@ public UserServiceImpl(UserRepo userRepo) {
...
+    public UserTO findById(long id) {
+        User user = userRepo.findById(id).orElseThrow(
+            () -> new NoSuchEntityException("User with id " + id + " not found"));
+        return ConvertToTO.userToTO(user);
     }
 
     @Override
-    public Iterable<User> findAll() {
-        return userRepo.findAll();
+    public Set<UserTO> findAll() {
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/implementation/UserServiceImpl.java:33]: Done

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java (hunk 8 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:21]: Done

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                .phone(client.getPhone())
+                .build();
+    }
+
+    public static ClinicTO clinicToTO(Clinic clinic) {
+        return ClinicTO.builder()
+                .clinicId(clinic.getId())
+                .link(clinic.getLink())
+                .name(clinic.getName())
+                .faculty(FacultyTO.builder()
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:49]: Done

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    public static ProjectTO projectToTO(Project project) {
+
+        Set<ClinicTO> clinics = project.getClinics().stream()
+                .map(clinic -> clinicToTO(clinic))
+                .collect(Collectors.toSet());
+
+        Set<CommitTO> commits = project.getCommits().stream()
+                .map(commit -> commitToTO(commit))
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:92]: Done

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+                .collect(Collectors.toSet());
+
+        Set<String> links = project.getLinks().stream()
+                .map(link -> link.getLink())
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:96]: Done

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        Set<String> requirementsForPerformers = project.getRequirementsForPerformers().stream()
+                .map(re -> re.getRequirement())
+                .collect(Collectors.toSet());
+
+        Set<StudentTO> students = project.getStudents().stream()
+                .map(student -> studentToTO(student))
+                .collect(Collectors.toSet());
+
+        Set<String> tags = project.getTags().stream()
+                .map(tag -> tag.getName())
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:108]: Done

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java (hunk 7 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:100]: Done

File: src/main/java/org/spburegistry/backend/controller/UserController.java (hunk 2 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/controller/UserController.java:23]: Idk, return to Iterable

File: src/main/java/org/spburegistry/backend/dto/ProjectTO.java (hunk 4 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/dto/ProjectTO.java:32]: Done

File: src/main/java/org/spburegistry/backend/entity/Clinic.java (hunk 5 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/entity/Clinic.java:33]: Ok

File: src/main/java/org/spburegistry/backend/entity/Project.java (hunk 6 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/entity/Project.java:68]: I think we need this later

File: src/main/java/org/spburegistry/backend/entity/Project.java [hunk 16]
```diff
@@ -6,87 +6,98 @@
...
+        @Nullable
+        private String resultLink;
+
+        @ManyToOne(fetch = FetchType.EAGER)
+        @JoinColumn(name = "client_id", nullable = false)
+        private Client client;
+
+        @EqualsAndHashCode.Exclude
+        @Builder.Default
+        @ManyToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/entity/Project.java:57]: Done

File: src/main/java/org/spburegistry/backend/entity/Project.java [hunk 17]
```diff
@@ -6,87 +6,98 @@
...
+        private Set<Student> students = new HashSet<>();
+
+        public void addStudent(Student student) {
+                students.add(student);
+                student.addProject(this);
+        }
+
+        @EqualsAndHashCode.Exclude
+        @Builder.Default
+        @ManyToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/entity/Project.java:72]: Done

File: src/main/java/org/spburegistry/backend/entity/Project.java [hunk 18]
```diff
@@ -6,87 +6,98 @@
...
+        @JoinTable(
+                name = "project_clinic", 
+                joinColumns = @JoinColumn(name = "project_id"), 
+                inverseJoinColumns = @JoinColumn(name = "clinic_id")
+        )
+        private Set<Clinic> clinics = new HashSet<>();
+
+        @EqualsAndHashCode.Exclude
+        @Builder.Default
+        @ManyToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/entity/Project.java:82]: Done

File: src/main/java/org/spburegistry/backend/entity/Student.java [hunk 19]
```diff
@@ -31,16 +32,22 @@ public class Student extends BaseEntity {
...
     private User user;
 
+    @Builder.Default
     @ManyToMany(mappedBy = "students")
-    private Set<Project> projects;
+    private Set<Project> projects = new HashSet<>();
 
+    public void addProject(Project project) {
+        projects.add(project);
+    }
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/entity/Student.java:44]: Ok

File: src/main/java/org/spburegistry/backend/entity/User.java [hunk 20]
```diff
@@ -24,11 +24,13 @@ public class User extends BaseEntity {
     @Enumerated(EnumType.STRING)
     private Role role;
 
-    @OneToOne(mappedBy = "user")
+    @EqualsAndHashCode.Exclude
+    @OneToOne(mappedBy = "user", cascade = CascadeType.ALL, optional = false, fetch = FetchType.LAZY)
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/entity/User.java:28]: Done

File: src/main/java/org/spburegistry/backend/service/implementation/ProjectServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+                        tag.setCreationTime(new Date());
+                        tag.setLastUpdateTime(new Date());
+                        tagRepo.save(tag);
+                    }
+                    return tag;
+                }).collect(Collectors.toSet());
+        newProject.setTags(tags);
+        client.addProject(newProject);
+        clinics.forEach(clinic -> clinic.addProject(newProject));
+        tags.forEach(tag -> tag.addProject(newProject));
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/implementation/ProjectServiceImpl.java:89]: Ok

File: src/main/java/org/spburegistry/backend/service/UserService.java (hunk 3 above)
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/UserService.java:8]: Done

File: src/main/java/org/spburegistry/backend/utils/ConvertToTO.java [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+        Set<String> links = project.getLinks().stream()
+                .map(link -> link.getLink())
+                .collect(Collectors.toSet());
+
+        Set<String> requirementsForPerformers = project.getRequirementsForPerformers().stream()
+                .map(re -> re.getRequirement())
+                .collect(Collectors.toSet());
+
+        Set<StudentTO> students = project.getStudents().stream()
+                .map(student -> studentToTO(student))
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertToTO.java:104]: Done

File: src/main/java/org/spburegistry/backend/repository/TagRepo.java [hunk 23]
```diff
@@ -4,5 +4,5 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface TagRepo extends JpaRepository<Tag, Long> {
-    
+    public Tag findByName(String name);
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/repository/TagRepo.java:7]: I think we can you this method in filter tag

File: src/main/java/org/spburegistry/backend/service/implementation/ClientServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+    public ClientTO addNewClient(ClientTO clientTO) {
+        Client client = clientRepo.getReferenceById(clientTO.getClientId());
+        if (client == null) {
+            client = Client.builder()
+                    .link(clientTO.getLink())
+                    .organizationName(clientTO.getOrgName())
+                    .phone(clientTO.getPhone())
+                    .build();
+            client.setCreationTime(new Date());
+            client.setLastUpdateTime(new Date());
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/implementation/ClientServiceImpl.java:37]: Ok

File: src/main/java/org/spburegistry/backend/service/implementation/ClientServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+            User user = User.builder()
+                    .name(clientTO.getName())
+                    .email(clientTO.getEmail())
+                    .build();
+            user.setCreationTime(new Date());
+            user.setLastUpdateTime(new Date());
+            user.setClient(client);
+            client.setUser(user);
+            clientRepo.save(client);
+            userRepo.save(user);
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/implementation/ClientServiceImpl.java:47]: Ok

File: src/main/java/org/spburegistry/backend/service/implementation/ProjectServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+    @Override
+    public ProjectTO findById(long id) {
+        Project project = projectRepo.findById(id).orElseThrow(
+                () -> new NoSuchEntityException("Project with id " + id + " not found"));
+        return ConvertToTO.projectToTO(project);
+
+    }
+
+    @Override
+    public ProjectTO addProject(ProjectTO projectTO) {
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/implementation/ProjectServiceImpl.java:53]: Done

File: src/main/java/org/spburegistry/backend/service/implementation/ClientServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+    @Autowired
+    public ClientServiceImpl(ClientRepo clientRepo, UserRepo userRepo) {
+        this.clientRepo = clientRepo;
+        this.userRepo = userRepo;
+    }
+
+    @Override
+    public ClientTO addNewClient(ClientTO clientTO) {
+        Client client = clientRepo.getReferenceById(clientTO.getClientId());
+        if (client == null) {
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/service/implementation/ClientServiceImpl.java:30]: I delete this file

File: src/main/java/org/spburegistry/backend/utils/ConvertFromTO.java [hunk 28]
```diff
@@ -0,0 +1,18 @@
...
+    public static Client clientFromTO(ClientTO clientTO) {
+        return Client.builder().link(null).build();
+            
+    }
+
+    public static Project projectFromTO(ProjectTO projectTO) {
+        return Project.builder()
+            .client(null).build();
+    }
+}
```
[the remark this answers is not in the document]
Comment (End-rey) [reply src/main/java/org/spburegistry/backend/utils/ConvertFromTO.java:18]: I delete this file

File: src/main/java/org/spburegistry/backend/dto/ProjectRequest.java [hunk 29]
```diff
@@ -0,0 +1,41 @@
+package org.spburegistry.backend.dto;
```
Comment (SerjVankovich) [inline src/main/java/org/spburegistry/backend/dto/ProjectRequest.java:1]: I think it is better to rename this class to ProjectRequestTO

Comment (End-rey) [reply src/main/java/org/spburegistry/backend/dto/ProjectRequest.java:1]: Done

**Review by SerjVankovich — APPROVED** [review]

**Review by MicAPic — APPROVED** [review]
